### PR TITLE
More app manager view tests

### PR DESCRIPTION
**Pull request by orangejenny**
Test coverage for issues like those fixed in
https://github.com/dimagi/commcare-hq/pull/6933
https://github.com/dimagi/commcare-hq/pull/7113
https://github.com/dimagi/commcare-hq/pull/7107

@sravfeyn

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/tests/test_views.py, line 110 - E303 too many blank lines (2)
* corehq/apps/app_manager/tests/test_views.py, line 122 - E201 whitespace after '{'
* corehq/apps/app_manager/tests/test_views.py, line 122 - E202 whitespace before '}'
* corehq/apps/app_manager/tests/test_views.py, line 156 - E201 whitespace after '{'
* corehq/apps/app_manager/tests/test_views.py, line 156 - E202 whitespace before '}'

**Comment by snopoke** [comment]
These tests only work if the app under test has all the features e.g. all the different module types. Also if we add something new later we'll need to update the test app.

**Comment by orangejenny** [comment]
> These tests only work if the app under test has all the features e.g. all the different module types.
I don't follow - the test app does contain all the features being tested.

> Also if we add something new later we'll need to update the test app.
Isn't that always true - new features mean updating/writing tests? (not being snarky)

**Comment by snopoke** [comment]
Writing new tests is easy but updating app json to make sure a new feature is tested is hard. I would say that its easier to have a test that runs through a bunch of different apps (each one with a feature to test). Then in future you can just add a new app json to the folder to test the new thing.

**Comment by snopoke** [comment]
Test failure: 

```
======================================================================
ERROR: setUpClass (corehq.apps.app_manager.tests.test_views.TestViews)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 117, in new_setUpClass
    result = original_setUpClass(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_views.py", line 22, in setUpClass
    cls.user = WebUser.create(cls.domain, cls.username, cls.password, is_active=True)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 2126, in create
    web_user.save()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1228, in save
    raise self.Inconsistent("CouchUser with username %s already exists" % self.username)
Inconsistent: CouchUser with username cornelius already exists
```

**Comment by orangejenny** [comment]
@snopoke Oh, I see what you mean about benefits of testing multiple apps instead of one big one. Will do.

File: corehq/apps/app_manager/tests/test_views.py [hunk 1]
```diff
@@ -88,3 +99,92 @@ def test_edit_commcare_profile(self):
         custom_properties = content["changed"]["custom_properties"]
 
         self.assertEqual(custom_properties["random"], "changed")
+
+    def _load_app(self, filename):
```
Comment (sravfeyn) [inline corehq/apps/app_manager/tests/test_views.py]: These utils methods are available in [`corehq.apps.app_manager.tests.util.TestFileMixin`](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/tests/util.py#L39) You could subclass `TestFileMixin` and reuse those same methods.

File: corehq/apps/app_manager/tests/test_views.py [hunk 2]
```diff
@@ -88,3 +99,92 @@ def test_edit_commcare_profile(self):
...
+    def test_reporting_module(self):
+        self._load_app('reporting_module.json')
+        self._test_status_codes(['view_module'], {
+            'domain': self.domain,
+            'app_id': self.app.id,
+            'module_id': 0,
+        })
+
+    def test_view_form(self):
+        self._load_app('basic_app.json')
```
Comment (sravfeyn) [inline corehq/apps/app_manager/tests/test_views.py]: (doesn't really concern this PR, but more of a general question about tests) For the ease of readability, explicitness and less changes in the app-source when the App-Model changes, one could argue to create the application on the fly using methods like `new_app`, and explicitly making it what you want using helper methods - but that could slow down the tests.

Wherever possible I prefer creating apps using the Application API methods and making it what you want vs loading it from a source file.
- Whenever app-model changes slightly, you would have to change all these source files - with API methods, you probably would have to change just those public methods
- Using API methods makes it explicit to the reader what a test expects the app be like.

But on the other end, loading from json files and avoiding database makes these tests fast. So I am not really sure what to choose, but I usually try to go with API methods. I am curious to listen what you and others think.

**Comment by sravfeyn** [comment]
Looks good apart from the [reusing util methods](https://github.com/dimagi/commcare-hq/pull/7195/files#r33021334) to load app from json files

File: corehq/apps/app_manager/tests/test_views.py (hunk 2 above)
Comment (snopoke) [inline corehq/apps/app_manager/tests/test_views.py]: +1 on creating the app using the API

On 23 June 2015 at 11:01, Sravan Reddy <notifications@github.com> wrote:

> In corehq/apps/app_manager/tests/test_views.py
> <https://github.com/dimagi/commcare-hq/pull/7195#discussion_r33022019>:
>
> > +        self._test_status_codes(['view_module'], {
> > +            'domain': self.domain,
> > +            'app_id': self.app.id,
> > +            'module_id': 0,
> > +        })
> > +
> > +    def test_reporting_module(self):
> > +        self._load_app('reporting_module.json')
> > +        self._test_status_codes(['view_module'], {
> > +            'domain': self.domain,
> > +            'app_id': self.app.id,
> > +            'module_id': 0,
> > +        })
> > +
> > +    def test_view_form(self):
> > +        self._load_app('basic_app.json')
>
> (doesn't really concern this PR, but more of a general question about
> tests) For the ease of readability, explicitness and less changes in the
> app-source when the App-Model changes, one could argue to create the
> application on the fly using methods like new_app, and explicitly making
> it what you want using helper methods - but that could slow down the tests.
>
> Wherever possible I prefer creating apps using the Application API methods
> and making it what you want vs loading it from a source file.
>
>    - Whenever app-model changes slightly, you would have to change all
>    these source files - with API methods, you probably would have to change
>    just those public methods
>    - Using API methods makes it explicit to the reader what a test
>    expects the app be like.
>
> But on the other end, loading from json files and avoiding database makes
> these tests fast. So I am not really sure what to choose, but I usually try
> to go with API methods. I am curious to listen what you and others think.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7195/files#r33022019>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

**Comment by benrudolph** [comment]
nice! this is great to see. i'm a big fan of having view tests just to sanity check things are working end to end

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/tests/test_views.py, line 6 - F401 'TestFileMixin' imported but unused
* corehq/apps/app_manager/tests/test_views.py, line 13 - F401 'import_app' imported but unused
* corehq/apps/app_manager/tests/test_views.py, line 13 - E501 line too long (120 > 115 characters)
* corehq/apps/app_manager/tests/test_views.py, line 153 - E303 too many blank lines (2)
* corehq/apps/app_manager/tests/test_views.py, line 174 - F841 local variable 'module' is assigned to but never used

**Comment by orangejenny** [comment]
@sravfeyn @snopoke updated to use Application API

**Pull request merged**